fedauth.go

```diff
@@ -0,0 +1,101 @@
+package mssql
+
+import (
+	"context"
+	"crypto/rsa"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/Azure/go-autorest/autorest/adal"
+	"golang.org/x/crypto/pkcs12"
+)
+
+const (
+	activeDirectoryEndpoint = "https://login.microsoftonline.com/"
+	azureSQLResource        = "https://database.windows.net/"
+	driverClientID          = "7f98cb04-cd1e-40df-9140-3bf7e2cea4db"
+)
+
+func fedAuthGetClientCertificate(clientCertPath, clientCertPassword string) (*x509.Certificate, *rsa.PrivateKey, error) {
+	pkcs, err := ioutil.ReadFile(clientCertPath)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to read the AD client certificate from path %s: %v", clientCertPath, err)
+	}
+
+	privateKey, certificate, err := pkcs12.Decode(pkcs, clientCertPassword)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Failed to read the AD client certificate from path %s: %v", clientCertPath, err)
+	}
+
+	rsaPrivateKey, isRsaKey := privateKey.(*rsa.PrivateKey)
+	if !isRsaKey {
+		return nil, nil, fmt.Errorf("AD client certificate at path %s must contain an RSA private key", clientCertPath)
+	}
+
+	return certificate, rsaPrivateKey, nil
+}
+
+func fedAuthGetAccessToken(ctx context.Context, resource, tenantID string, p connectParams, log optionalLogger) (accessToken string, err error) {
+	// The activeDirectoryEndpoint URL is used as a base against which the
+	// tenant ID is resolved. When the workflow provides a complete endpoint
+	// URL for the tenant, the URL resolution just returns that endpoint.
+	oauthConfig, err := adal.NewOAuthConfig(activeDirectoryEndpoint, tenantID)
+	if err != nil {
+		log.Printf("Failed to obtain OAuth configuration for endpoint %s and tenant %s: %v", activeDirectoryEndpoint, tenantID, err)
```

Comment:
You must not log like this.
Rather, you can do return "", fmt.Errorf("Failed to obtain... %v", err)
but no log to console.

Reply:
The call to log.Printf was being directed to the driver logger - not the console, but I changed it to a wrapped error return.